### DISCUSSION TITLE
forward gradient to mi.Transform4f

### DISCUSSION THREAD

**Discussion by zichenwang01** [mitsuba-renderer/mitsuba3#965] [Q&A] [answered]
Hi,

I have a single parameter and want to forward it to a mi.Transform4f instance. Right now I am not getting the correct gradients and would like to ask if I did anything wrong. Thanks!

I am running in a Linux environment on a 2080ti GPU.

```
import drjit as dr
import mitsuba as mi

mi.set_variant('cuda_ad_rgb')

p = mi.Float(1e-3)
dr.enable_grad(p)

transform = mi.Matrix4f([ [1, 0, 0, 0],
                          [0, 1, 0, p],
                          [0, 0, 1, 0],
                          [0, 0, 0, 1] ])
transform = mi.Transform4f(transform)
dr.enable_grad(transform)

dr.set_grad(p, 1)
# dr.forward(p)
dr.forward_to(transform)
print(dr.grad(transform)) # prints all 0
```

**Comment by wjakob** [comment]
The second `enable_grad` looks wrong to me, what is the purpose of that?

**Reply by zichenwang01** [reply]
I thought it might help to make sure that transform is gradient enabled. (Deleting that does not affect the result)

**Reply by ziyi-zhang** [reply, marked as answer]
```
import drjit as dr
import mitsuba as mi

mi.set_variant('cuda_ad_rgb')

p = mi.Float(0)
dr.enable_grad(p)

transform = mi.Transform4f.translate(mi.Vector3f(0, p, 0))

dr.forward(p, dr.ADFlag.ClearEdges)

print(dr.grad(transform.matrix))
print(dr.grad(transform.inverse_transpose))
```
This returns
```
[[[0.0, 0.0, 0.0, 0.0],
  [0.0, 0.0, 0.0, 1.0],
  [0.0, 0.0, 0.0, 0.0],
  [0.0, 0.0, 0.0, 0.0]]]
[[[0.0, 0.0, 0.0, 0.0],
  [0.0, 0.0, 0.0, 0.0],
  [0.0, 0.0, 0.0, 0.0],
  [0.0, -1.0, 0.0, 0.0]]]
```

`mi.transform` is a struct that store both the matrix and the inverse transpose. In the computational graph, the inverse transpose is computed from the matrix itself.
`dr.forward` with the default flag cleans everything -- in the above example `transform.matrix` would have gradient zero since it is an intermediate node that has been cleared during traversal.

**Reply by zichenwang01** [reply]
Setting the ADFlag helped solve the problem. Many thanks!
